### PR TITLE
Cache: sync committee index position

### PR DESCRIPTION
**Pull request by terencechain**
# Description

This PR brings in the sync committee index cache from #9123. Sync committee index cache is a fifo cache that caches sync committee index position across current and next sync committee period. This will be greatly used for computing sync committee subnet, position index with regards to subnet and more

# Other info

Ratio on `devnet2` over the last 24hrs
Screen Shot 2021-08-02 at 12 03 07 PM

### CODE REVIEW

File: beacon-chain/cache/sync_committee.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	pos, err := s.idxPositionInCommittee(root, valIdx)
+	if err != nil {
+		return nil, err
+	}
+	if pos == nil {
+		return []types.CommitteeIndex{}, nil
+	}
```
Comment (prestonvanloon) [inline beacon-chain/cache/sync_committee.go:92]: This block is identical to the contents of CurrentPeriodIndexPosition. Consider refactoring these to share the same code.

File: beacon-chain/cache/sync_committee.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	maxSyncCommitteeSize = uint64(3) // Allows 3 forks to happen around `EPOCHS_PER_SYNC_COMMITTEE_PERIOD` boundary.
```
Comment (rauljordan) [inline beacon-chain/cache/sync_committee.go:17]: Is this better to track in the global config? Might be hard to find in the future as it has no mention of "forks", it just says maxSyncCommitteeSize

File: beacon-chain/cache/sync_committee.go (hunk 1 above)
Comment (rauljordan) [reply beacon-chain/cache/sync_committee.go:92]: still pending resolution @terencechain

File: beacon-chain/cache/sync_committee.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+		return nil, err
+	}
+	if pos == nil {
+		return []types.CommitteeIndex{}, nil
+	}
+	return pos.nextPeriod, nil
+}
+
+// Helper function for `CurrentPeriodIndexPosition` and `NextPeriodIndexPosition` to return a mapping
+// of validator index to its index(s) position in the sync committee.
```
Comment (rauljordan) [inline beacon-chain/cache/sync_committee.go:97]: Comment needed that this is not thread safe and caller needs to acquire lock

File: beacon-chain/cache/sync_committee.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+		} else {
+			positionsMap[validatorIndex].currentPeriod = append(positionsMap[validatorIndex].currentPeriod, types.CommitteeIndex(i))
+		}
+	}
+
+	nsc, err := st.NextSyncCommittee()
+	if err != nil {
+		return err
+	}
+	for i, pubkey := range nsc.Pubkeys {
```
Comment (rauljordan) [inline beacon-chain/cache/sync_committee.go:149]: Some duplication in these two for loops that could be refactored perhaps

File: beacon-chain/cache/sync_committee.go (hunk 1 above)
Comment (terencechain) [reply beacon-chain/cache/sync_committee.go:92]: offline discussion there's no easy way to refactor this. "A little duplication is far cheaper than the wrong abstraction"

https://discord.com/channels/476244492043812875/774005284188323870/872154033362726932

File: beacon-chain/cache/sync_committee.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/cache/sync_committee.go:17]: No, it's a cache specific value. I dont think we would put cache max size in global config. The comment means the max tolerance is 3 forks around sync committee period as that's where cache gets refreshed

File: beacon-chain/cache/sync_committee.go (hunk 4 above)
Comment (terencechain) [reply beacon-chain/cache/sync_committee.go:149]: Can't refactor this, one is current period and one is next period. The structure is very different

Comment (rauljordan) [reply beacon-chain/cache/sync_committee.go:149]: it's doable, but it would be a weird abstraction you would need to do with callbacks..not worth it

**Review by rauljordan — APPROVED** [review]

**Pull request merged**